test ParameterHandler::MultipleSelection

tjhei opened the pull request.

tjhei commented:
I wonder if we should allow things like "set v=bla,bla" like we currently do. Thoughts?

bangerth commented:
You mean you want to disallow repeating the same element more than once? What if your case is
```
  set Temperature values on the different boundaries   = 4293, 4293, 1723, 1723
```
? It makes perfect sense to repeat the same value multiple times.

Or maybe I just don't understand your question?

tjhei commented:
Turns out some weird things happen: 1) the input from the parameter file will replace double spaces by a single space (even inside parameters!). See updated test.

bangerth commented on include/deal.II/base/parameter_handler.h:
```diff
@@ -595,20 +595,20 @@ namespace Patterns
-   * constructor.
+   * constructor argument. The input is allowed to be empty or contain values
+   * more than once. An arbitrary number of spaces
+   *
```
inkomplet sentence.

tjhei commented:
> It makes perfect sense to repeat the same value multiple times.

But this pattern is a selection of predefined names (and your example doesn't make sense because you would need to add every single integer to the definition of the pattern). Or am I wrong?

bangerth commented:
That's slightly weird. I would have expected that it only trims spaces at the front and back, but not in the middle.

bangerth commented:
> But this pattern is a selection of predefined names (and your example doesn't
> make sense because you would need to add every single integer to the
> definition of the pattern). Or am I wrong?

But surely we can come up with other cases where one would have to repeat some 
symbolic name multiple times. I don't think this to be a particularly weird case.

I'm not opposed to adding a boolean argument to the class that indicates 
whether duplicates are allowed or not.

tjhei commented:
> But surely we can come up with other cases where one would have to repeat some symbolic name multiple times. I don't think this to be a particularly weird case.

From the name "Selection" I was assuming that a value is either selected or not. If you think it is okay, we can leave it as is. (it is documented now)

tjhei commented:
I found the problem, fixing...

tjhei commented:
Updated with the fixes to ParameterHandler. This PR is getting more complicated than anticipated...

bangerth commented on include/deal.II/base/parameter_handler.h:
```diff
@@ -595,20 +595,19 @@ namespace Patterns
+   *
+   * For example, if the string to the constructor was <tt>"ucd|gmv|eps"</tt>,
+   * then the following would be legal input: <tt>eps</tt>, <tt>gmv,eps</tt>,
+   * or <tt></tt>.
```
The last one of these isn't going to print very well :-) How about enclosing each of these legal inputs in quotes?

bangerth commented on tests/bits/parameter_handler_21.cc:
```diff
@@ -0,0 +1,74 @@
+  check("  input 2  ,  have spaces  ","have spaces|input 2","");
+
+  // weird: two spaces in the input get reduces to one
+  check("input 2","have spaces|input 2","set v=   input 2  ,   have spaces  ");
```
so is the comment still true? I don't see a case in the input that actually has two spaces, other than around the `=` and `,` characters.

tjhei replied:
this is now fixed.

bangerth commented:
Looks pretty good.

Do you want to extend `trim` to also throw away tabs at the beginning and end? (The current code uses `std::isspace` to determine whether a character is a space or tab, but on looking up what the function does, `std::isblank` actually seems like a better fit.)

tjhei commented:
> Do you want to extend trim to also throw away tabs at the beginning and end? 

The first thing we do when parsing a line in ParameterHandler is to replace \t by ' ' (thinking about it this means a value can not have a tab value right now :-) ). If you want I can add \t to trim(), otherwise I would leave it as is for now.

I think I addressed all your comments.

tamiko commented:
This looks fine.

bangerth commented:
I think the issue of tabs in `trim()` is separate from this here, so let's not worry too much.
